content/posts/numpy/generator-rng/index.md

```diff
@@ -0,0 +1,304 @@
+---
+title: "Random numbers done right"
+date: 2022-04-25
+draft: false
+description: "
+Use `np.random.Generator` and don't seed! A quick explainer on random number
+generation with NumPy.
+"
+tags: ["NumPy", "tutorials", "random-numbers"]
+displayInList: true
+author: ["Pamphile Roy"]
+
+---
+
+Do you need random numbers? You use NumPy? NumPy has a canonical way to
+produce random numbers with
+[`np.random.Generator`](https://numpy.org/doc/stable/reference/random/index.html).
+There is a say in Python, "global variables are bad". And you guessed correctly,
+it's also bad to use `np.random.random()` or other generators that rely on a
+global state! I will also demonstrate why you should refrain yourself from
+using a seed.
+
+In the following, we assume that _NumPy_, _SciPy_ and _Matplotlib_ are
+installed and imported:
+
+```python
+import numpy as np
+import scipy.stats as stats
+import matplotlib.pyplot as plt
+```
+
+## Random done wrong
+
+```python
+# the following is wrong and only exists for demonstration purposes!
+np.random.seed(123456)
+sample = np.random.random(1_024)
+```
+
+Aoutch! Two lines of codes and both are wrong. This is a simple snippet to
+generate 1024 points using NumPy's random number generator that is very common
+in codes and tutorials online. Sadly/surprisingly even on famous blogs.
+
+There are 2 problems I want to talk about:
+
+1. [global state](#global-id),
+2. [seed](#seed-id).
+
+## Goodbye global state {#global-id}
+
+Behind `np.random.random` a number generator is used. This generator has an
+internal state which is _global_. This state is shared and as new points/samples
+are requested, this state is used and updated. The state can be set with a
+_seed_ value.
+
+As with any global variable, it means that it can be changed by other
+subprocesses, threads, other libraries, etc. Take the following code. It uses
+4 threads to call `np.random.random` 10 times and then prints the last value.
+We execute this 5 times to see how the value changes.
+
+```python
+# the following is wrong and only exists for demonstration purposes!
+import concurrent.futures
+
+for _ in range(5):
+    np.random.seed(123456)  # fix the state of the global generator
+
+    sample = []
+    with concurrent.futures.ThreadPoolExecutor(4) as executor:
+        futures = [executor.submit(np.random.random, 1) for _ in range(10)]
+        for future in concurrent.futures.as_completed(futures):
+            sample.append(future.result())
+
+    print(sample[-1])
+
+# [0.33622174]
+# [0.5430262]
+# [0.96671784]
+# [0.37674972]
+# [0.96671784]
```

Comment:
Mildly curious but you literally get the same value here, which shows the danger of using global in anything other than vanilla single threaded/process code.

Reply:
It's not that much of a surprise. It's getting reseeded to the same state in the main thread at the top of each iteration. If it were seeded only once before the loop and we still got repeats, that _would_ be more of a problem, but the design of even `RandomState` doesn't allow that problem (thanks to the GIL and Cythonized code).

This is something of a contrived example. The main problem with threads and the global in real problems is the loss of determinism, but much of the article is arguing against determinism. If one were committed to seedlessness, there really wouldn't be a problem with using the `np.random` functions and threads even with the hidden global.